In [1]:
import pandas as pd
import numpy as np

### 1.1.1

In [2]:
day = 365
hours = 24
minutes = 60
seconds = 60
secondsYear = day * hours * minutes * seconds
secondsYear

31536000

### 1.1.2

In [3]:
age = 20
secondsAge = age * secondsYear
secondsAge

630720000

### 1.1.3

In [4]:
older = secondsYear > 700000000
older

False

### 1.1.4

In [5]:
p = 0.4
result1 = p*(1-p)
print(result1)
a = 3
x_0 = 2
x = 2.5
result2 = a/x_0*(1+x/x_0)**(-a-1)
print(result2)

0.24
0.058527663465935076


### 1.2.1

In [6]:
movies = ["movie1", "movie2", "movie3", "movie4", "movie5", "movie6"]

### 1.2.2

In [7]:
first_three_movies = movies[:3]
first_three_movies

['movie1', 'movie2', 'movie3']

### 1.2.3

In [8]:
first_two_movies = movies[:2]
last_two_movies = movies[-2:]
f2_l2 = first_two_movies + last_two_movies
f2_l2


['movie1', 'movie2', 'movie5', 'movie6']

### 1.2.4

In [9]:
movie_reverse = movies[::-1]
movie_reverse

['movie6', 'movie5', 'movie4', 'movie3', 'movie2', 'movie1']

### 1.3.1

In [10]:
numbers = []
for i in range(0, 10):
    numbers.append(i+70)
numbers

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]

### 1.3.2

In [11]:
sum = 0
for i in range(0, len(numbers)):
    sum+=numbers[i]
result = sum/len(numbers)
result

74.5

### 1.3.3

In [12]:
sum100 = 0
for i in range(1, 101):
    sum100+=i
sum100

5050

### 1.3.4

In [13]:
names = ["Adam", "Ashin", "Inukai", "Tanaka", "Ikki"]
seats = [33, 12, 45, 2, 17]
for i in range(0, len(names)):
    print(names[i] + ":", seats[i])

Adam: 33
Ashin: 12
Inukai: 45
Tanaka: 2
Ikki: 17


### 1.3.5

In [14]:
my_dist = {}
l = [1, 3, 5, 7, 11, 13, 17, 19]
for i in l:
    my_dist[i] = i**2
my_dist


{1: 1, 3: 9, 5: 25, 7: 49, 11: 121, 13: 169, 17: 289, 19: 361}

### 1.3.6

In [15]:
sumDist = 0
for value in my_dist.values():
    sumDist += value
sumDist

1024

### 1.4.1

In [16]:
def time(num):
    minutes = num % 100
    hours = num // 100
    return hours + minutes / 60
print(time(1015))
print(time(345))


10.25
3.75


### 2.1.1

In [17]:
df = pd.read_csv("nycflights13.csv.bz2", sep="\t")

#### 2.1.2.a

In [18]:
df.shape[1]

19

There are 19 variables

#### 2.1.2.b

In [19]:
df.shape[0]

336776

There are 336776 rows or observations

#### 2.1.2.c

In [20]:
df.head(5)

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T10:00:00Z  
1      1416     5      29  2013-01-01T10:00:00Z  
2      1089     5      40  2013-01-01T10:00:00Z  
3      1576     5      45  2013-01-01T10:00:00Z  
4       762     6       0  2013-01-01T11:00:00Z

From the first few lines only, I would say it looks reasonable. There are no missings, all the value lies in the appropriate range.

#### 2.2.1

In [21]:
df['origin'].nunique()

3

3 NYC airports included

### 2.2.2

In [22]:
df['dest'].nunique()

105

There are 105 unique airports from NYC

### 2.2.3

In [23]:
df[df['dest'] == 'SEA'].shape[0]

3923

There are 3923 flights from NYC to Seattle

### 2.2.4

In [24]:
df[df['dest'] == 'GEG'].shape[0]

0

No, there aren't any flight from NYC to Spokane

### 2.2.5

In [25]:
np.mean(df['arr_delay'])

6.89537675731489

The average delay is about 7 minutes. I focused on arrival delay because. Since in this case, it would affect whether my boss makes it to the meeting on time. Since no specific destination was given, I calculated the "typical" delay by taking the mean of all arrival delays from New York.

### 2.2.6

In [26]:
print(df['arr_delay'].isna().sum())
print(df['arr_delay'].max())
print(df['arr_delay'].min())

9430
1272.0
-86.0


There are 9430 NAs. The min delay seems reasonable as airplane can arrive one hour and a half early. The max delay is a bit suspecious which is a almost one day delay. But it may happen due to weird weather conditions and external factors.

### 2.3.1

In [27]:
df_aus = df[df.dest == "AUS"].copy()
df_aus.shape[0]

2439

In 2013, 2439 flights were from NYC to Austin

### 2.3.2

In [28]:
df_aus['carrier'].nunique()

6

6 airlines fly from NYC to Austin

### 2.3.3

In [29]:
df_aus['carrier'].value_counts()

carrier
B6    747
UA    670
AA    365
DL    357
WN    298
9E      2
Name: count, dtype: int64

### 2.3.4

In [30]:
df_aus[df_aus['arr_delay'] > 15].shape[0] / df_aus.shape[0] * 100

24.641246412464124

24.6 % of flights to Austin were delayed at arrival by more than 15 min

### 2.3.5

In [31]:
df_aus['flight_duration'] = time(df_aus["arr_time"]) - time(df_aus["dep_time"])

### 2.3.6

In [32]:
min_duration = df_aus['flight_duration'].min()
print(min_duration)
df_aus[df_aus['flight_duration'] == min_duration]

-21.75


year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
246678  2013      6   26    2323.0            2051      152.0     138.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
246678            2355      103.0      UA    1615  N35407    EWR  AUS   

        air_time  distance  hour  minute             time_hour  \
246678     181.0      1504    20      51  2013-06-27T00:00:00Z   

        flight_duration  
246678           -21.75

The min value is -21.75 which is not reasonable as the plane can't fly negative time. This may happen when the plane departs at late night (in this case 23:23) and arrive after midnight (in this case 1:38)

### 2.3.7

In [33]:
df_aus['flight_duration'] = np.where(
    df_aus['flight_duration'] < 0,
    df_aus['flight_duration'] + 24,
    df_aus['flight_duration']
)
df_aus['flight_duration'].min()

2.2333333333333343

### 2.3.8

In [34]:
df_aus[(df_aus['origin'] == 'JFK')].head(5)
df_aus[(df_aus['origin'] == 'JFK')].flight_duration.mean()

3.1150777320530403

It's due to different time zone. Since departure time and arrival time are recorded based on local time zone. New York is using GMT-4 and Austin is using GMT-5. There is a 1 hour difference between them. Based on my calculation the average is 3.11 and the one listed on the booking website is around 4.05. So the 1 hour difference matches the actual time difference.

### 2.4.1

My main concern was the quality of the data. Since there are quite a lot missing values in critical columns like arrival time and departure time. These will affect my calculation on duration and find patterns between scheduled time and actual fligh time. My method of taking mean might not be proper as well. I use it for convenience, but I should also do better data cleaning and more statistical analysis to make sure my result is plausible.

### 2.4.2

A travel company could use this data to predict which airline or routes has most or less flight delay happen and suggest to book or not book this flight. This is quite important since some users need to hurry up for urgent matters or transfers, and flight delays have a great impact on them.

### 2.4.3

Yes. When analyzing delays, I might put some airline company to "most delay" or something without considering external factors such as weather conditions and congestions in the airport. If my result is posted to public, it might cause unfair negative impact on some companies.